File: src/main/java/com/formacraft/common/model/composite/CompositeSpec.java
package com.formacraft.common.model.composite;

import com.formacraft.common.model.build.BuildingSpec;
import com.formacraft.common.model.path.PathSpec;

import java.util.List;

/**
 * 复合结构规格
 * 包含多个子结构及其相对坐标
 * 用于生成城市、要塞、村庄等复合建筑
 */
public class CompositeSpec {
    
    /**
     * 子结构定义
     */
    public static class SubStructure {
        private String type;  // TOWER, HOUSE, BRIDGE, WALL, etc.
        private BuildingSpec spec;
        private Offset offset;

        public SubStructure() {}

        public String getType() {
            return type;
        }

        public void setType(String type) {
            this.type = type;
        }

        public BuildingSpec getSpec() {
            return spec;
        }

        public void setSpec(BuildingSpec spec) {
            this.spec = spec;
        }

        public Offset getOffset() {
            return offset;
        }

        public void setOffset(Offset offset) {
            this.offset = offset;
        }
    }

    /**
     * 三维坐标偏移
     */
    public static class Offset {
        public int x;
        public int y;
        public int z;

        public Offset() {}

        public Offset(int x, int y, int z) {
            this.x = x;
            this.y = y;
            this.z = z;
        }
    }

    private List<SubStructure> structures;
    private List<PathSpec> paths;  // 路径列表

    public CompositeSpec() {}

    public List<SubStructure> getStructures() {
        return structures;
    }

    public void setStructures(List<SubStructure> structures) {
        this.structures = structures;
    }

    public List<PathSpec> getPaths() {
        return paths;
    }

    public void setPaths(List<PathSpec> paths) {
        this.paths = paths;
    }
}

File: python_backend/app/models/composite_spec.py
from pydantic import BaseModel
from typing import List, Optional
from .building_spec import BuildingSpec


class Vec3i(BaseModel):
    """三维坐标"""
    x: int
    y: int
    z: int


class SubStructure(BaseModel):
    """子结构定义"""
    type: str  # TOWER, HOUSE, BRIDGE, WALL, etc.
    spec: BuildingSpec  # 完整的 BuildingSpec
    offset: Vec3i  # 相对于 Composite 的 Origin 的偏移


class PathSpec(BaseModel):
    """路径规格"""
    id: str = ""  # 唯一标识符，如 "main_road", "path_1"
    from_pos: Vec3i  # 起点（使用 from_pos 避免 Python 关键字冲突）
    to_pos: Vec3i    # 终点
    width: int = 3   # 道路宽度（默认 3 格）
    material: str = "minecraft:gravel"  # 道路材质
    style: str = "default"  # 道路样式：default / curved / stepped / decorated


class CompositeSpec(BaseModel):
    """复合结构规格（BuildingSpec 的集合 + 相对坐标 + 路径）"""
    structures: List[SubStructure]
    paths: Optional[List[PathSpec]] = []  # 路径列表（可选）

